### PR TITLE
Min - LB125.LC_tot

### PR DESCRIPTION
**Pull request by chenminbnu**
@bpbond , I created a new pull request.  Maybe you could delete the old one. thanks.

#262

### CODE REVIEW

File: R/zchunk_LB125.LC_tot.R [hunk 1]
```diff
@@ -39,65 +42,131 @@ module_aglu_LB125.LC_tot_DISABLED <- function(command, ...) {
...
+    L125.LC_bm2_R_Yh_GLU %>%
+      arrange(GCAM_region_ID, GLU, year) %>%
+      mutate(changing_rate = value / lag(value)) %>%          # calculate the changing rate
+      replace_na(list(changing_rate = 1)) %>%                 # assign na to 1 (no change)
+      ungroup() %>%
+      select(changing_rate) ->
+      LC_check
+
+    # Stop if the changing rate out of the tolerance boundaries
+    if (max(abs(LC_check - 1)) > LAND_TOLERANCE)
```
Comment (chenminbnu) [inline R/zchunk_LB125.LC_tot.R]: Time_shift test keeps reporting "ERROR" here and throwing out the error message in Line 93

File: R/zchunk_LB125.LC_tot.R [hunk 2]
```diff
@@ -1,26 +1,29 @@
 #' module_aglu_LB125.LC_tot
 #'
-#' Briefly describe what this chunk does.
+#' Build Total Land Cover by GCAM region, or by GCAM region and GLU, or by GCAM region/GLU/year.
 #'
```
Comment (abigailsnyder) [inline R/zchunk_LB125.LC_tot.R]: maybe use and instead of or, since all three tables are outputs

File: R/constants.R [hunk 3]
```diff
@@ -178,3 +178,10 @@ emissions.EDGAR_YEARS <- 1971:2008
 emissions.EPA_MACC_YEAR <- 2030  # Must be either 2020 or 2030
 emissions.MAC_TAXES <- c( 0, 5, 10, 15, 32, 66, 129, 243, 486, 1093 ) # Range of costs in 1990 USD
 emissions.CONV_C_CO2 <- 44 / 12 # Convert Carbon to CO2
+
+#======================================================================
+# Land area aggregation constants
+
+LAND_TOLERANCE <- 0.001
```
Comment (bpbond) [inline R/constants.R]: Probably put these in the `aglu` section in this file

File: R/zchunk_LB125.LC_tot.R [hunk 4]
```diff
@@ -39,65 +42,128 @@ module_aglu_LB125.LC_tot_DISABLED <- function(command, ...) {
...
+      replace_na(list(changing_rate = 1)) %>%                 # assign na to 1 (no change)
+      ungroup() %>%
+      select(changing_rate) ->
+      LC_check
+
+    # Stop if the changing rate out of the tolerance boundaries
+    if (max(abs(LC_check - 1)) > LAND_TOLERANCE)
+    {stop("ERROR: Interannual fluctuation in global land cover exceeds tolerance threshold")}
+
+    #Write out the totals, by region and by region x GLU
```
Comment (bpbond) [inline R/zchunk_LB125.LC_tot.R]: Space after `#` please

File: R/zchunk_LB125.LC_tot.R [hunk 5]
```diff
@@ -39,65 +42,128 @@ module_aglu_LB125.LC_tot_DISABLED <- function(command, ...) {
...
+    L125.LC_bm2_R_Yh_GLU %>%
+      filter(year == min(year)) %>%                        # using the starting year only
+      group_by(GCAM_region_ID, GLU) %>%                    # group by GCAM_region_ID and GLU
+      summarise(value = sum(value)) %>%                    # summarize area by region and GLU
+      rename(LC_bm2 = value) ->                            # rename the column
```
Comment (bpbond) [inline R/zchunk_LB125.LC_tot.R]: For what it's worth, this isn't (IMHO) a helpful comment. We have to assume a certain amount of R literacy of the reader, and `rename(LC_bm2 = value)` is clear and obvious.

Comment (chenminbnu) [inline R/zchunk_LB125.LC_tot.R]: @bpbond I've revised it according to your comments, but syncing is failed. Do you have a sense of why?

Comment (bpbond) [inline R/zchunk_LB125.LC_tot.R]: Probably because I resolved conflicts in my commit below. If you're around, happy to help resolve this. -B

**Comment by bpbond** [comment]
Holding on review by @kvcalvin

**Comment by bpbond** [comment]
This is now failing the (new) timeshift test because of an upstream problem. Specifically, from 2005 to 2006 the sum drops precipitously:
```
Browse[2]> LC_check[85:90,]
# A tibble: 11 × 5
   GCAM_region_ID    GLU  year     value changing_rate
            <int>  <chr> <int>     <dbl>         <dbl>
1               1 GLU023  2003 145.91388     0.9999999
2               1 GLU023  2004 145.91387     0.9999999
3               1 GLU023  2005 145.91385     0.9999999
4               1 GLU023  2006  75.88945     0.5200976     <----- here
5               1 GLU023  2007  76.33387     1.0058562
6               1 GLU023  2008  76.77830     1.0058221
```

...due to missing rows in the un-summed dataset:

```
Browse[2]> L125.LC_bm2_R_LT_Yh_GLU %>% filter(year %in% 2005:2006, GLU=="GLU023") %>% arrange(year, Land_Type) %>% print(n=50)
# A tibble: 34 × 5
   GCAM_region_ID        Land_Type    GLU  year        value
            <int>            <chr>  <chr> <int>        <dbl>
1               1           Forest GLU023  2005   4.66102739
2               8           Forest GLU023  2005  15.39014387
3               1        Grassland GLU023  2005  18.96798000
4               8        Grassland GLU023  2005  46.51878000
5               1     HarvCropLand GLU023  2005  66.33259357    <--- missing in 2006
6               8     HarvCropLand GLU023  2005 208.85246939    <--- missing in 2006
7               1  OtherArableLand GLU023  2005   4.13623643    <--- missing in 2006
8               8  OtherArableLand GLU023  2005 161.33360061    <--- missing in 2006
9               1          Pasture GLU023  2005   0.10216865
10              8          Pasture GLU023  2005  15.22101157
11              8    RockIceDesert GLU023  2005   0.17942000
12              8        Shrubland GLU023  2005   0.71103000
13              8           Tundra GLU023  2005   5.57690000
14              1  UnmanagedForest GLU023  2005  29.12492261
15              8  UnmanagedForest GLU023  2005 356.39638613
16              1 UnmanagedPasture GLU023  2005  22.23612135
17              8 UnmanagedPasture GLU023  2005  74.50497843
18              1        UrbanLand GLU023  2005   0.35280000
19              8        UrbanLand GLU023  2005   1.09961000
20              1           Forest GLU023  2006   4.52525979
21              8           Forest GLU023  2006  14.76504656
22              1        Grassland GLU023  2006  19.25367000
23              8        Grassland GLU023  2006  51.09128200
24              1          Pasture GLU023  2006   0.08394656
25              8          Pasture GLU023  2006  15.36784339
26              8    RockIceDesert GLU023  2006   0.17942000
27              8        Shrubland GLU023  2006   0.72296000
28              8           Tundra GLU023  2006   5.58196600
29              1  UnmanagedForest GLU023  2006  29.28867621
30              8  UnmanagedForest GLU023  2006 358.18681344
31              1 UnmanagedPasture GLU023  2006  22.37746944
32              8 UnmanagedPasture GLU023  2006  73.83162261
33              1        UrbanLand GLU023  2006   0.36042400
34              8        UrbanLand GLU023  2006   1.12964000
```

At first I thought that `LB122.LC_R_Cropland_Yh_GLU` had a problem, but, much to @abigailsnyder 's relief, no, it's operating correctly; everything ELSE is wrong. In particular, this LB125 chunk needs to filter its `temp-data-inject/` inputs so they are limited to `HISTORICAL_YEARS` (because under the timeshift test this is no longer true); and `LB120.LC_GIS_R_LTgis_Yh_GLU` needs to filter its UrbanLand, Tundra, and RockIceDesert outputs similarly.

**Comment by bpbond** [comment]
I have fixed the above problem but this is still failing, it looks like because of a problem at the _beginning_ of the time series, this time. Here's what things look like when running the timeshift:

```
Browse[2]> LC_check[1190:1195,]
# A tibble: 6 × 5
  GCAM_region_ID    GLU  year    value changing_rate
           <int>  <chr> <int>    <dbl>         <dbl>
1              3 GLU086  2005 523.3504     1.0000000
2              3 GLU087  1971 252.2930     1.0000000
3              3 GLU087  1972 252.7305     1.0017341    <---- greater than 0.001
4              3 GLU087  1973 253.0798     1.0013819
5              3 GLU087  1974 253.1496     1.0002760
6              3 GLU087  1975 253.1365     0.9999483

Browse[2]> L125.LC_bm2_R_LT_Yh_GLU %>% filter(year %in% 1971:1972, GLU=="GLU087", GCAM_region_ID==3) %>% arrange(year, Land_Type) %>% spread(year, value) %>% mutate(diff = `1972` - `1971`) %>% print(n=50) 
# A tibble: 38 × 6
   GCAM_region_ID        Land_Type    GLU       `1971`       `1972`          diff
            <int>            <chr>  <chr>        <dbl>        <dbl>         <dbl>
11              3           Forest GLU087 1.250900e-01 1.288600e-01  0.0037700000
12              3        Grassland GLU087 6.456768e+00 6.584836e+00  0.1280680000
13              3     HarvCropLand GLU087 1.818568e+01 1.838505e+01  0.1993682240
14              3  OtherArableLand GLU087 8.766514e+00 8.651094e+00 -0.1154202743
15              3    RockIceDesert GLU087 2.163721e+02 2.165524e+02  0.1803190000
16              3        Shrubland GLU087 1.184742e+00 1.193674e+00  0.0089320000
17              3  UnmanagedForest GLU087 0.000000e+00 0.000000e+00  0.0000000000
18              3        UrbanLand GLU087 1.202172e+00 1.234644e+00  0.0324720000
```

When running with normal `HISTORICAL_YEARS`, this looks like
```
  GCAM_region_ID       Land_Type    GLU     `1971`     `1972`       diff
           <int>           <chr>  <chr>      <dbl>      <dbl>      <dbl>
1              3          Forest GLU087   0.125090   0.128860  0.0037700
2              3       Grassland GLU087   6.456768   6.584836  0.1280680
3              3    HarvCropLand GLU087  19.081556  18.830327 -0.2512285   <--- different
4              3 OtherArableLand GLU087   7.772305   7.669975 -0.1023305
5              3   RockIceDesert GLU087 216.372059 216.552378  0.1803190
6              3       Shrubland GLU087   1.184742   1.193674  0.0089320
7              3 UnmanagedForest GLU087   0.000000   0.000000  0.0000000
8              3       UrbanLand GLU087   1.202172   1.234644  0.0324720
```

...which brings us back to you again @abigailsnyder  😈

**Comment by abigailsnyder** [comment]
@bpbond I took a little look at the L122 files. I particularly looked at the old vs new agreement - for OtherArableLand, the largest magnitude difference between the old and the new DS is 10e-8 (with sumtests to 10e-7). For HarvCropLand, the magnitudes are the same.

Filtering to GLU087 in region 3 for both L122 tables from the old AND new ds, the numbers that you get for 1971 and 1972 are identical to those you've output above (to 6 decimals). 

I did bring the two L122 files over from the old DS as temp-data-inject just to see if it changes anything. The time shift error is resolved (though two of the L125 outputs don't match the old DS still).  So that tells me it's again something that's happening out several decimal places. Maybe related to the division step (value / lag(value)) in forming LC_check? where that 7th decimal place difference might matter?

**Comment by bpbond** [comment]
Just to clarify, if you use `temp-data-inject` L122 files as inputs to L125, the timeshift error goes away?

**Comment by abigailsnyder** [comment]
@bpbond yes, if I bring the L122 files over from the old DS to the temp-data-inject folder, update all the references to be temp-data-inject, and then move the L122 files to be included in the gather in the first computation pipeline, the timeshift error goes away and it's just the L125 output mismatch errors remaining. Would you like me to push my code?

**Comment by abigailsnyder** [comment]
since the old and new ds L122 files only differ by 10e-8 or so, that makes me think it's the division step. But I could be absolutely biased by the L100/L107 PR we just had where that was the issue.

**Comment by bpbond** [comment]
Huh! No, that's fine, thanks. OK, time to bring in the big man I guess. 

@pkyle we're getting larger than `LAND_TOLERANCE` swings in global land cover when running the timeshift test--i.e., when changing the data system's historical/future settings. Variations are up to 0.0072, while the max allowed is 0.001.
```
 changing_rate   
 Min.   :0.9979  
 1st Qu.:1.0000  
 Median :1.0000  
 Mean   :1.0000  
 3rd Qu.:1.0000  
 Max.   :1.0072  
```

Is there anything magic about the 0.001 value currently use? I ask because one quick-and-dirty fix would be loosen the tolerance to 0.01, for example.

**Comment by bpbond** [comment]
@abigailsnyder I guess I don't get why that division step would be OK normally, but trigger this failure when we limit `HISTORICAL_YEARS` to 1971-2005.

**Comment by abigailsnyder** [comment]
@bpbond  for me it wasn't even limiting to all vs historic years. It's old L122 vs new L122. 

I tested changing line 86 to `mutate(changing_rate = round(value, 5) / lag(round(value, 5)))`. 

With the new L122 files (not temp-data-inject of the old, in other words), there's still a time shift error. So I guess it might not be the division step. The old and new L122 files agree to the 7th/8th decimal place, so I'm not sure what else could be magnifying that small difference

**Comment by abigailsnyder** [comment]
* the old DS numbers get through limiting the historical years whereas the new DS numbers somehow don't. And the step is always ok normally, regardless of which numbers are used.

**Comment by bpbond** [comment]
@abigailsnyder can we touch base about this tomorrow morning, if you're in and available? Thanks.

**Comment by abigailsnyder** [comment]
sounds good, I'll be around

**Comment by pkyle** [comment]
Sorry I wasn't monitoring this discussion today. I don't know that there is a reason for using 0.001 as opposed to 0.01, but at some point the model won't allow it. It will throw an error message and crash. I'd imagine I selected a value higher than the tolerance threshold, because really I was using this step to test the methods of land data processing, to make sure each place has the same total land cover in all periods. Without knowing where this 0.007 difference is coming from, it's hard to say whether this is totally trivial, or actually indicative of a leak in the method. And I don't know off the top of my head whether GCAM will crash on an inter-annual difference in total land cover of 0.007 for any region.

**Comment by bpbond** [comment]
Thanks @abigailsnyder for resetting us to a known point, where LB125 runs and passes its oldnew test, but fails the timeshift (partly for the reasons above, but also we think because of rounding issues). I need to set this aside for the next 48 hours but will return to it at the end of the week.

**Comment by abigailsnyder** [comment]
I ran a bunch of tests but I can't quite synthesize a fix from them. @bpbond I can go over them with you on Thursday, since that may just be faster and easier than trying to explain in writing.

**Comment by bpbond** [comment]
Picking this up again. At the failure point, under timeshift conditions in LB125, we have this:

```R
Browse[2]> L125.LC_bm2_R_LT_Yh_GLU %>%
+          arrange(GCAM_region_ID, GLU, year) %>% filter(GCAM_region_ID==1, GLU=="GLU023", year %in% 2004:2008) %>% spread(year, value)
# A tibble: 8 × 8
  GCAM_region_ID        Land_Type    GLU     `2004`     `2005`      `2006`     `2007`    `2008`
*          <int>            <chr>  <chr>      <dbl>      <dbl>       <dbl>      <dbl>     <dbl>
1              1           Forest GLU023  4.7023009  4.6610274  4.52525979  4.2627171  3.970372
2              1        Grassland GLU023 18.3579080 18.9679800 19.25367000 19.5393600 19.825050
3              1     HarvCropLand GLU023 65.8162726 66.3325936          NA         NA        NA
4              1  OtherArableLand GLU023  5.5072294  4.1362364          NA         NA        NA
5              1          Pasture GLU023  0.1931612  0.1021687  0.08394656  0.1074641  0.110053
6              1  UnmanagedForest GLU023 29.0219851 29.1249226 29.28867621 29.5792049 29.899536
7              1 UnmanagedPasture GLU023 21.9689088 22.2361213 22.37746944 22.4770779 22.597615
8              1        UrbanLand GLU023  0.3461000  0.3528000  0.36042400  0.3680480  0.375672
```

@abigailsnyder if you're around today let's talk in person?

**Comment by bpbond** [comment]
Further notes:

## Normal run

```R
Browse[2]> LC_check %>% arrange(GCAM_region_ID, GLU, year) %>% filter(GCAM_region_ID==3, GLU=="GLU087", year %in% 1971:1974)
Source: local data frame [4 x 6]
Groups: GCAM_region_ID, GLU [1]

  GCAM_region_ID    GLU  year    value change_rate change
           <int>  <chr> <int>    <dbl>       <dbl>  <dbl>
1              3 GLU087  1971 252.1947          NA     NA
2              3 GLU087  1972 252.1947           1  2e-06
3              3 GLU087  1973 252.1947           1  2e-06
4              3 GLU087  1974 252.1947           1  2e-06
Browse[2]> L125.LC_bm2_R_LT_Yh_GLU %>%
+     arrange(GCAM_region_ID, GLU, year) %>% filter(GCAM_region_ID==3, GLU=="GLU087", year %in% 1971:1974) %>% spread(year, value)
# A tibble: 8 × 7
  GCAM_region_ID       Land_Type    GLU     `1971`     `1972`     `1973`     `1974`
*          <int>           <chr>  <chr>      <dbl>      <dbl>      <dbl>      <dbl>
1              3          Forest GLU087   0.125090   0.128860   0.132630   0.136400
2              3       Grassland GLU087   6.456768   6.584836   6.712904   6.518308
3              3    HarvCropLand GLU087  19.081556  18.830327  18.579099  18.707678
4              3 OtherArableLand GLU087   7.772305   7.669975   7.567644   7.465314
5              3   RockIceDesert GLU087 216.372059 216.552378 216.732697 216.913016
6              3       Shrubland GLU087   1.184742   1.193674   1.202606   1.154394
7              3 UnmanagedForest GLU087   0.000000   0.000000   0.000000   0.000000
8              3       UrbanLand GLU087   1.202172   1.234644   1.267116   1.299588
```

## Timeshift run

```R
Browse[3]> LC_check %>% arrange(GCAM_region_ID, GLU, year) %>% filter(GCAM_region_ID==3, GLU=="GLU087", year %in% 1971:1974)
Source: local data frame [4 x 6]
Groups: GCAM_region_ID, GLU [1]

  GCAM_region_ID    GLU  year    value change_rate     change
           <int>  <chr> <int>    <dbl>       <dbl>      <dbl>
1              3 GLU087  1971 252.2930          NA         NA
2              3 GLU087  1972 252.7305    1.001734 0.43750895
3              3 GLU087  1973 253.0798    1.001382 0.34924147
4              3 GLU087  1974 253.1496    1.000276 0.06985945


L125.LC_bm2_R_LT_Yh_GLU %>%
               arrange(GCAM_region_ID, GLU, year) %>% filter(GCAM_region_ID==3, GLU=="GLU087", year %in% 1971:1974) %>% spread(year, value)
# A tibble: 8 × 7
  GCAM_region_ID       Land_Type    GLU     `1971`     `1972`     `1973`     `1974`
*          <int>           <chr>  <chr>      <dbl>      <dbl>      <dbl>      <dbl>
1              3          Forest GLU087   0.125090   0.128860   0.132630   0.136400
2              3       Grassland GLU087   6.456768   6.584836   6.712904   6.518308
3              3    HarvCropLand GLU087  18.185682  18.385051  18.496151  18.707678
4              3 OtherArableLand GLU087   8.766514   8.651094   8.535674   8.420253
5              3   RockIceDesert GLU087 216.372059 216.552378 216.732697 216.913016
6              3       Shrubland GLU087   1.184742   1.193674   1.202606   1.154394
7              3 UnmanagedForest GLU087   0.000000   0.000000   0.000000   0.000000
8              3       UrbanLand GLU087   1.202172   1.234644   1.267116   1.299588
```

Why are the `HarvCropLand` and `OtherArableLand` numbers changing between these runs?

**Comment by bpbond** [comment]
OK, as @abigailsnyder points out, this is almost certainly happening because of line 115 in `LB122.LC_R_Cropland_Yh_GLU`:

```R
    fallowland_year <- max(AGLU_HISTORICAL_YEARS)
    # ...and then lots of things happen using this variable
```

**Comment by bpbond** [comment]
I realized that the timeshift test isn't testing what we want, actually. What we care about is the ability to adjust `HISTORICAL_YEARS` and `FUTURE_YEARS`, basically, parameters that control what data GCAM treats as past versus present. It doesn't (and shouldn't) need to change things like `AGLU_HISTORICAL_YEARS` because this relates to _how we process this particular dataset_. These two things are confounded across much of the data system (see, again, #455 as well).

Do you agree @kvcalvin ?

Scaling back timeshift so that it's adjusting only HY and FY, this chunk still fails timeshift, but only a few rows, and by much less:
```R
Browse[2]> print(na.omit(LC_check[out,]))
Source: local data frame [6 x 6]
Groups: GCAM_region_ID, GLU [4]

  GCAM_region_ID    GLU  year     value change_rate       change
           <int>  <chr> <int>     <dbl>       <dbl>        <dbl>
1             23 GLU047  2010 99.396990   0.9987827 -0.121139655
2             26 GLU212  2006  2.087764   0.9953063 -0.009845584
3             27 GLU116  2008 16.374261   0.9978403 -0.035440748
4             27 GLU116  2009 16.319409   0.9966502 -0.054851189
5             27 GLU116  2010 16.280361   0.9976072 -0.039048439
6             31 GLU207  2006  1.613142   0.9951093 -0.007928109
```

I think at this point we have two choices: (i) loosen the tolerance from 0.001 to 0.005, or (ii) decree that this tolerance check doesn't occur when under timeshift.

**Comment by bpbond** [comment]
OK once the Travis tests pass I think this is ready to merge. Thanks all. 🎉

**Comment by pkyle** [comment]
Yeah I think there was a slight misunderstanding--in the old data system, I used the objects "model_base_years" and "model_future_years" to mean the years for GCAM, whereas the "historical_years" and "future_years" were just for processing data. Those were separated for a number of reasons (e.g., hindcasting runs, only running a small subset of the total available historical years, probably others too). The only reason we'd ever change the historical_years is if we got more data going further forward or back in history. I certainly never intended for people to come in and e.g. set historical_years to 1971:2006 and have that produce a similar or even functional dataset; I just never would have built it for the situation where we get less data over time, or where the max(historical_years) is anything less than 2010.
Going forward, it's fine to maintain that capacity, but we will probably have to re-set a number of objects to a hard-wired 2010 year where the code currently tries to pull the latest historical year.

**Comment by bpbond** [comment]
@pkyle moving your comment to #480

**Pull request merged**